src/EFCore.SqlServer/Query/Internal/SqlServerSqlTranslatingExpressionVisitor.cs

```diff
@@ -199,6 +208,63 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
             return translation3;
         }
 
+        // Translate string.Join to CONCAT_WS
+        if (method == StringJoinMethodInfo
+            && methodCallExpression.Arguments[1] is NewArrayExpression newArrayExpression
+            && _sqlServerCompatibilityLevel >= 140)
+        {
+            if (TranslationFailed(methodCallExpression.Arguments[0], Visit(methodCallExpression.Arguments[0]), out var delimiter))
+            {
+                return QueryCompilationContext.NotTranslatedExpression;
+            }
+
+            var arguments = new SqlExpression[newArrayExpression.Expressions.Count + 1];
+            arguments[0] = delimiter!;
+
+            var isUnicode = delimiter!.TypeMapping?.IsUnicode == true;
+            var stringTypeMapping = _typeMappingSource.FindMapping(typeof(string));
+
+            for (var i = 0; i < newArrayExpression.Expressions.Count; i++)
+            {
+                var argument = newArrayExpression.Expressions[i];
+                if (TranslationFailed(argument, Visit(argument), out var sqlArgument))
+                {
+                    return QueryCompilationContext.NotTranslatedExpression;
+                }
+
+                // CONCAT_WS returns a type with a length that varies based on actual inputs (i.e. the sum of all argument lengths, plus
+                // the length needed for the delimiters). We don't know column values (or even parameter values, so we always return max.
+                // We do vary return varchar(max) or nvarchar(max) based on whether we saw any nvarchar mapping.
+                if (sqlArgument!.TypeMapping?.IsUnicode == true)
+                {
+                    isUnicode = true;
+                }
+
+                // CONCAT_WS filters out nulls, but string.Join treats them as empty strings; coalesce unless we know we have a non-nullable
+                // argument.
+                arguments[i + 1] = sqlArgument switch
+                {
+                    ColumnExpression { IsNullable: false } => sqlArgument,
+                    SqlConstantExpression constantExpression => constantExpression.Value is null
+                        ? new SqlConstantExpression(string.Empty, stringTypeMapping)
+                        : constantExpression,
+                    _ => Dependencies.SqlExpressionFactory.Coalesce(
+                        sqlArgument,
+                        Dependencies.SqlExpressionFactory.Constant(string.Empty, typeof(string)))
+                };
```

Comment:
I would like to eventually ensure that the nullability processor can simplify `COALESCE` by dropping everything after the first known-non-null subexpression (and all of the known-null subexpressions). This would make it possible to simply use 
```suggestion
                arguments[i + 1] = Dependencies.SqlExpressionFactory.Coalesce(
                        sqlArgument,
                        Dependencies.SqlExpressionFactory.Constant(string.Empty, typeof(string)));
```
here

Reply:
@ranma42 thanks, that's an interesting direction... A couple of thoughts:

* First, we can **partially** do this simplification at the source, i.e. improve SqlExpressionFactory.Coalesce() to not actually add the coalesce node for non-nullable columns/constants - I'm making that change in this PR. This has the small advantage of keeping the tree a bit cleaner, i.e. no unneeded coalesce node before the nullability processor.
* But you're right that for the general case of arbitrary expressions, this would (currently) need to happen at the nullability processor, since that's the only place where we know the nullability of arbitrary expressions (anything beyond column/constant).
* This general design is something we've discussed in the past; I've opened #33889 with some thoughts, but that's a very long-term, high-level architecture question that we can't really tackle in the near term.

Reply:
I agree that there is value in keeping the expression tree as clean as possible; OTOH duplicating the null propagation/simplification logic across the codebase has its disadvantages.
I think that an interesting approach to manage some simple cases (that are nonetheless currently repeated/spread across several files) could be a wrapper (possibly even as a collection of extension methods) that abstracts the pattern you are writing and makes it easily re-usable in several places.
I am thinking about something like `CoalesceAndSimplify` (or other similar factory methods) that pre-emptively handles trivial optimizations (local&cheap ones). For an example of another compiler/expression translator that does this, see LLVM and how it performs [trivial constant folding](https://llvm.org/docs/tutorial/MyFirstLanguageFrontend/LangImpl04.html#trivial-constant-folding) while building its AST.

A nit/question: is there a reason why one of the empty strings is built as `new SqlConstantExpression(string.Empty, stringTypeMapping)` and the other one as `Dependencies.SqlExpressionFactory.Constant(string.Empty, typeof(string))`?